### PR TITLE
APEXMALHAR-2430 Optimizing Join accumulation

### PR DESCRIPTION
**Pull request by KapoorHitesh**
@chinmaykolhatkar Please review

### CODE REVIEW

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java [hunk 1]
```diff
@@ -121,136 +120,109 @@ private void createSetters()
...
       }
     }
+    addNonMatchingRightStream(rightStream, rightGettersStream, result);
     return result;
   }
 
-  public abstract void addNonMatchingResult(List<Map<String, Object>> result, Map<String, Object> requiredMap, Set nullFields);
-
-  public abstract int getLeftStreamIndex();
+  public abstract void addNonMatchingResult(Collection<Object> left, Map<String,PojoUtils.Getter> leftGettersStream, List<Object> result);
```
Comment (chinmaykolhatkar) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java]: Can you add a javadoc comment for this containing, what is expected to happen here and when will this be called?

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java [hunk 2]
```diff
@@ -121,136 +120,109 @@ private void createSetters()
...
 
+  public abstract void addNonMatchingRightStream(Multimap<List<Object>, Object> rightStream, Map<String,PojoUtils.Getter> rightGettersStream, List<Object> result);
```
Comment (chinmaykolhatkar) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java]: Can you add a javadoc comment for this containing, what is expected to happen here and when will this be called?

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java [hunk 3]
```diff
@@ -121,136 +120,109 @@ private void createSetters()
...
-      }
-    }
-    for (String field : map2.keySet()) {
-      if (!keySetStream2.contains(field)) {
-        map1.put(field, map2.get(field));
-      }
-    }
-    return map1;
-  }
+  public abstract int getLeftStreamIndex();
```
Comment (chinmaykolhatkar) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java]: Can you add a javadoc comment for this containing, what is expected to happen here and when will this be called?

Comment (chinmaykolhatkar) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java]: Can this be made protected?

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java (hunk 2 above)
Comment (chinmaykolhatkar) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java]: Can this be made protected?

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java (hunk 1 above)
Comment (chinmaykolhatkar) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java]: Can this be made protected?

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java [hunk 4]
```diff
@@ -121,136 +120,109 @@ private void createSetters()
...
       }
-      out.add(o);
     }
 
-    return out;
   }
 
-
-  public List<Map<String, Object>> getAllCombo(int streamIndex, List<List<Map<String, Object>>> accu, Map<String, Object> curMap)
+  public List<Object> getAllCombo(List<Multimap<List<Object>, Object>> accu)
```
Comment (chinmaykolhatkar) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java]: For the sake of readers of this accumulation, can you describe the logic in this method and how it handles all case of joins?

**Review by chinmaykolhatkar — CHANGES_REQUESTED** [review]
Functionally looks good. Some minor comments. Can you please take care of them, squash and rebase..?

**Comment by KapoorHitesh** [comment]
@chinmaykolhatkar have made the requested changes.

**Pull request merged**